pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -50,67 +52,113 @@ public Operator<IntermediateResultsBlock> run() {
-   *  This is nlogk operation which can be quite expensive for a large n.
-   *  In the above example, if the docs in the segment are already sorted by col1 and col2 then there is no need for
-   *  sorting at all (only limit is needed).
-   * @return true is all columns in order by clause are sorted . False otherwise
+   * This functions returns the number of expressions that are sorted by the implicit order in the index.
+   *
+   * This means that query that uses these expressions in its order by doesn't actually need to sort from 0 to the
+   * given number (excluded) of expressions, as they are returned in the correct order.
+   *
+   * This method supports ASC and DESC order and ensures that all prefix expressions follow the same order. For example,
+   * ORDER BY sorted_col1 ASC, sorted_col2 ASC and ORDER BY sorted_col1 DESC, sorted_col2 DESC will return 2 but
+   * ORDER BY sorted_col1 DESC, sorted_col2 ASC and ORDER BY sorted_col1 ASC, sorted_col2 DESC will return 1 while
+   * ORDER BY not_sorted, sorted_col1 will return 0 because the first column is not sorted.
+   *
+   * It doesn't make sense to add literal expressions in an order by expression, but if they are included, they are
+   * considered sorted and its ASC/DESC is ignored.
+   *
+   * @return the max number that guarantees that from the first expression to the returned number, the index is already
+   * sorted.
    */
-  private boolean isAllOrderByColumnsSorted(List<OrderByExpressionContext> orderByExpressions) {
-    for (OrderByExpressionContext orderByExpression : orderByExpressions) {
-      if (!(orderByExpression.getExpression().getType() == ExpressionContext.Type.IDENTIFIER)
-          || !orderByExpression.isAsc()) {
-        return false;
+  private int getSortedColumnsPrefix(List<OrderByExpressionContext> orderByExpressions) {
+    boolean asc = orderByExpressions.get(0).isAsc();
+    for (int i = 0; i < orderByExpressions.size(); i++) {
+      if (!isSorted(orderByExpressions.get(i), asc)) {
+        return i;
       }
-      String column = orderByExpression.getExpression().getIdentifier();
-      if (!_indexSegment.getDataSource(column).getDataSourceMetadata().isSorted()) {
+    }
+    // If we reach here, all are sorted
+    return orderByExpressions.size();
+  }
+
+  private boolean isSorted(OrderByExpressionContext orderByExpression, boolean asc) {
+    switch (orderByExpression.getExpression().getType()) {
+      case LITERAL: {
+        return true;
+      }
+      case IDENTIFIER: {
+        if (!orderByExpression.isAsc() == asc) {
+          return false;
+        }
+        String column = orderByExpression.getExpression().getIdentifier();
+        return _indexSegment.getDataSource(column).getDataSourceMetadata().isSorted();
+      }
+      case FUNCTION: // we could optimize monotonically increasing functions
+      default: {
         return false;
       }
     }
-    return true;
+  }
+
+  private boolean isOrderByOptimizationEnabled() {
```

Comment:
I guess you mean to move it to `QueryOptionsUtils`. I'll do that.

Anyway, the `SelectionPlanNode.isOrderByOptimizationEnabled` is still there because it is free and I think 
```java
if (sortedColumnsPrefixSize > 0 && isOrderByOptimizationEnabled()) {
```

is more readable than

```java
if (sortedColumnsPrefixSize > 0 && QueryOptionsUtils.isOrderByOptimizationEnabled(_queryContext.getQueryOptions()))  {
```

Reply:
What I meant is that since this is on by default, we can choose to not have it (always on).

If you feel it is useful to have the ability to turn it off, we can put it in the `QueryOptionsUtils` (which you already did), and store the key in the constant for easier tracking. For the existing naming convention, suggest changing it to `skipOrderByOptimization` and make it default to `false`